src/rgw/rgw_sal_dbstore.h

```diff
@@ -259,6 +260,130 @@ class DBNotification : public Notification {
     }
   };
 
+  /*
+   * For multipart upload, below is the process flow -
+   *
+   * MultipartUpload::Init - create head object of the original object (if already doesn't
+   *                    exist) & meta obj (src_obj_name + "." + upload_id)
+   *                     [ Meta object stores all the parts upload info]
+   * MultipartWriter::process - create all data/tail objects with obj_name same as
+   *                        meta obj (so that they can all be identified & deleted 
+   *                        during abort)
+   * MultipartUpload::Abort - Just delete meta obj .. that will indirectly delete all the
+   *                     uploads associated with that upload id / meta obj so far.
+   * MultipartUpload::Complete - Rename all data/tail objects to orig object name and update
+   *                         metadata of the orig object.
```

Comment:
is it really necessary to do renames here? it seems like we should just be able to write the 'head object' here with its links to the data already intact

i guess that means that `ObjectName` shouldn't be used directly as a key to the `ObjectDataTable`?

i think the rados equivalent of this key would be the oid_prefix (see `RGWObjManifest::set_prefix()`), which is a random string that we generate for each upload. giving each upload a separate key is crucial to supporting racing writes to the same object

when an upload completes, we can write that key/oid_prefix to the head object so it knows where to find the data. and an overwrite of that object could update that key atomically to point at the data of a new upload

Reply:
What is a head object in dbstore.  Shouldn't this be clearly identified as an "upload?"

Reply:
For DBStore, so far at least, there was no need to maintain manifest or links to the data in 'ObjectDataTable'. Having ObjectName as a foreign key has taken care of overwrite, deletion and rename with a single db query (on ObjectTable).

By having separate OID to represent object in ObjectDataTable may result in explicit querying and updating the objects based on OID for every operation.

I still haven't checked or tested the race conditions in detail. But to begin with (first pass), the idea is to use sqlite transaction lock or if really needed to use simple mutex for every op.

Reply:
> What is a head object in dbstore. Shouldn't this be clearly identified as an "upload?"

Similar to Rados, in DBStore, an Object HEAD which contains, metadata + first chunk (default 1K now) of data is stored in Object table and rest of the data (tail object/s) is stored in ObjectData table. Have used same terms to easy reference.

Reply:
multipart uploads aside, how do normal object uploads work with racing writers?

how do they write data to the ObjectDataTable before the head object is created?

Reply:
The current implementation is to create head object (if not existing already) in ObjectTable during Object::AtomicWriter::prepare(), write tail data to ObjectDataTable in Object::AtomicWriter::process() followed by updating the head object metadata & first chunk in Object::AtomicWriter::complete().

But I got your point. This current method works only for sequential writes. 

To understand s3 object upload concurrency semantics better, I have added some notes and possible results (highlighted the optimal ones) at - https://docs.google.com/document/d/1xCoHT5DCujqbe1pnEfYpSaSqiBcW8z87CgQCC5LivOo/edit#heading=h.nvrx5mhkcz1r  (section: Object concurrent access semantics)

Please share your feed back on how the s3 server should respond (also if multiple results are possible ). thanks!